src/Build.UnitTests/ChangeWaves_Tests.cs

```diff
@@ -0,0 +1,232 @@
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
+
+                TransientTestFile file = env.CreateFile("proj.csproj", projectFile);
+
+                ProjectCollection collection = new ProjectCollection();
+                MockLogger log = new MockLogger();
+                collection.RegisterLogger(log);
+
+                collection.LoadProject(file.Path).Build().ShouldBeTrue();
+                log.AssertLogContains("Hello World!");
+                ChangeWaves.DisabledWave = null;
+            }
+            BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
+        }
+
+        [Theory]
+        [InlineData("test")]
+        [InlineData("    ")]
+        [InlineData("")]
+        [InlineData("16-7")]
+        [InlineData("16x7")]
+        [InlineData("16=7")]
+        public void InvalidCallerForIsFeatureEnabledThrows(string waveToCheck)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", "16.8");
+                Shouldly.Should.Throw<InternalErrorException>(() => ChangeWaves.IsChangeWaveEnabled(waveToCheck));
+            }
+            BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
+        }
+
+        [Theory]
+        [InlineData("test", "16.8")]
+        [InlineData("16_8", "5.7")]
+        [InlineData("16x8", "20.4")]
+        [InlineData("garbage", "18.20")]
+        public void InvalidFormatThrowsWarningAndLeavesFeaturesEnabled(string disableFromWave, string featureWave)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", disableFromWave);
+                ChangeWaves.IsChangeWaveEnabled(featureWave).ShouldBe(true);
+
+                string projectFile = @"
+                    <Project>
+                        <Target Name='HelloWorld' Condition=""'$(MSBUILDDISABLEFEATURESFROMVERSION)' == '" + ChangeWaves.EnableAllFeatures + @"' and $([MSBuild]::IsChangeWaveEnabled('" + featureWave + @"'))"">
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
+
+                TransientTestFile file = env.CreateFile("proj.csproj", projectFile);
+
+                ProjectCollection collection = new ProjectCollection();
+                MockLogger log = new MockLogger();
+                collection.RegisterLogger(log);
+
+                Project p = collection.LoadProject(file.Path);
+                p.Build().ShouldBeTrue();
+
+                log.AssertLogContains("invalid format");
+                log.AssertLogContains("Hello World!");
+                ChangeWaves.DisabledWave = null;
+            }
+            BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
```

Comment:
This is resetting the change wave environment variable at a time when it isn't set to anything. You need to put it right after `env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", disableFromWave);` for it to affect this test.

Reply:
If it resets the environment variable, wouldn't you want that to happen at the end of the test? I can see a potential need for least moving it before the testenvironment gets disposed.

Reply:
I'd do both, but if you don't reset it to anything that will interfere with other tests, I think just the beginning is ok. The bigger point is that if you reset it at the end, it will reset to nothing, whereas you want it to reset to the value you provided.

(For clarity, I should note that I really mean taking the value of the environment variable into account anew in Traits, not actually resetting its value, but they're equivalent in this case in practice.)

Reply:
I hadn't considered it effectively _applying_ the environment variable. It didn't work calling it after setting the environment variable, I'll try resetting it at both times as well.